app/main.py

```diff
@@ -4,31 +1,172 @@
+                self.is_drowned = True
+                for key in self.decks:
+                    self.decks[key] = self.decks[key].next()
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(
+            self,
+            ships: list[tuple[tuple[int, int], ...]],
+            title_word: str | None = "РЕСПУБЛІКА"
+    ) -> None:
+        self.title_word = title_word
+        self.fields = Battleship.create_fields(ships)
+        self._validate_field()
+
+    @staticmethod
+    def create_fields(
+            ships: list[tuple[tuple[int, int], ...]]
+    ) -> dict[tuple, Ship]:
+        fields = {}
+        for ship in ships:
+            ship_object = Ship(*ship)
+            for cell in ship_object.decks:
+                fields[cell] = ship_object
+        return fields
+
+    def fire(self, location: tuple[int, int]) -> str:
+        if location in self.fields:
+            ship = self.fields[location]
+            ship.fire(*location)
+            if ship.is_drowned:
+                return "Sunk!"
+            return "Hit!"
+        return "Miss!"
+
+    def column_coordinates_string(self) -> str:
+        if self.title_word is None or len(self.title_word) != 10:
+            return "   " + " ".join(str(i) + " " for i in range(10)) + "\n"
+        return "   " + " ".join(char + " " for char in self.title_word) + "\n"
+
+    def print_field(self) -> None:
+        field_str = self.column_coordinates_string()
+
+        for row in range(10):
+            field_str += str(row) + " "
+            for col in range(10):
+                cell = (row, col)
+                if cell in self.fields:
+                    field_str += self.fields[cell].get_deck(*cell).value
+                else:
+                    field_str += Tile.EMPTY.value
+            field_str += "\n"
+        print(field_str)
+
+    @staticmethod
+    def _validate_ship_types(ships: set) -> None:
+        ship_deck_counts = Counter(len(ship) for ship in ships)
+
+        required_counts = {1: 4, 2: 3, 3: 2, 4: 1}
+
+        for ship_size, required_count in required_counts.items():
+            if ship_deck_counts.get(ship_size, 0) != required_count:
+                raise ShipTypesCountError(
+                    f"Field must have {required_count} {ship_size}-deck ships"
+                )
+
+    def _validate_ship_placement(self, ships: set) -> None:
+        for ship in ships:
+            ship_neighbours = ship.get_adjacent_cells()
```

Comment:
The `ship` variable here is expected to be an instance of the `Ship` class, but it seems to be a tuple of coordinates. Ensure that `ships` is a set of `Ship` objects, not tuples, when calling `_validate_ship_placement`.